### PR TITLE
docs: Basic version of optimizers guide

### PR DESCRIPTION
**Pull request by dexter2206**
## Description

Include description of feature this PR introduces or a bug that it fixes. Include the following information:

- Context: why is it needed? In case of bug, what was the cause for the bug?
- Concise description of the implemented solution.
- If any dependencies are added, list them and justify why they are needed.

## Please verify that you have completed the following steps

- [ ] I have self-reviewed my code.
- [ ] I have included test cases validating introduced feature/fix.
- [ ] I have updated documentation.

### CODE REVIEW

File: docs/guides/optimizers.rst [hunk 1]
```diff
@@ -4,7 +4,398 @@
...
+======================
+This guide will introduce you to everything you need to know to optimize scalar functions using
+`orquestra-opt` package. In particular we'll discuss:
+
+- What functions can be minimized? Definition of cost function.
+- Gradients, and how to add them to your cost function.
+- Orquestra's approach for implementing optimization algorithms - the
+  :class:`Optimizer <orquestra.opt.api.Optimizer>` interface.
+- How to record history of evaluation of the cost function.
+- List of optimizers are currently available.
```
Comment (mstechly) [inline docs/guides/optimizers.rst:27]:
```suggestion
- List of currently available optimizers.
```

File: docs/guides/optimizers.rst [hunk 2]
```diff
@@ -4,7 +4,398 @@
...
+
+Adding gradients
+================
+Some optimization methods use gradients to move across the optimization landscape in the
+correct direction. You can enrich your cost function with gradient in several ways.
+
+To start with, cost function with gradients are just normal cost functions having additional
+`gradient` attribute. This attribute should also be a callable returning gradient
+of you function at a given point. Hence, one possible way of defining a cost function with
+gradient is to just write a class with `__call__` dunder method and a gradient attribute.
```
Comment (mstechly) [inline docs/guides/optimizers.rst:61]:
```suggestion
gradient is to just write a class with `__call__` method and a gradient attribute.
```

I'd drop `dunder` from the text, might be confusing if someone doesn't know what it is (and I guess a lot of people won't)

File: docs/guides/optimizers.rst [hunk 3]
```diff
@@ -4,7 +4,398 @@
...
+.. code-block:: text
+
+        history: []
+           nfev: None
+            nit: 1000
+     opt_params: array([ 5.11528491e-21, -5.11528491e-21,  5.38660015e-21])
+      opt_value: 8.134774054521418e-41
+
+
+As we can see, once constructed, :class:`ScipyOptimizer <orquestra.opt.optimizers.ScipyOptimizer>`
```
Comment (mstechly) [inline docs/guides/optimizers.rst:143]: Maybe worth mentioning somewhere that the output of minimize is adopted from `scipy`?

File: docs/guides/optimizers.rst [hunk 4]
```diff
@@ -4,7 +4,398 @@
...
+   parameters: [0.77104168 0.88808436 0.82558708], order: [0 2 1]
+   parameters: [0.45159923 0.5376592  0.14075176], order: [2 0 1]
+   parameters: [0.93415374 0.78783696 0.35753713], order: [2 1 0]
+
+We see that if our function stores artifacts, the entries stored in recorder's :code:`history`
+attribute are of :class:`HistoryEntryWithArtifacts <orquestra.opt.history.recorder.HistoryEntryWithArtifacts>`
+type, which has additional field called :code:`artifacts`. This field stores a dictionary of the
+artifacts that your cost function saved for given evaluation.
+
+Of course, there's no need for you cost function to store the same artifacts for each evaluation.
```
Comment (mstechly) [inline docs/guides/optimizers.rst:318]:
```suggestion
Of course, there's no need for your cost function to store the same artifacts for each evaluation.
```

File: docs/guides/optimizers.rst [hunk 5]
```diff
@@ -4,7 +4,398 @@
...
+  :start-after: # custom optimizer
+  :end-before: # --- End
+
+
+List of currently available optimizers
+======================================
+
+We conclude this guide by listing all optimizers currently available in orquestra-opt.
+
+Optimizers available with the default installation:
```
Comment (mstechly) [inline docs/guides/optimizers.rst:388]: Two thing:
1. To improve searchability, I'd also add what optimizers are available through Scipy, Qiskit, etc., at least the most popular ones.
2. Vlad added PSO recently, might be worth adding it to the list?

**Comment by mstechly** [comment]
Also: we had a placeholder for both `Optimizers` and `Cost function` docs, you've filled in only `Optimizers`.
I think that it makes sense to just get rid of the CF guide, but you'll need to scan through docs and remove any references to it.

**Comment by dexter2206** [comment]
> Also: we had a placeholder for both `Optimizers` and `Cost function` docs, you've filled in only `Optimizers`. I think that it makes sense to just get rid of the CF guide, but you'll need to scan through docs and remove any references to it.

I think I properly deleted it in [dcb9ddd](https://github.com/zapatacomputing/orquestra-core/pull/68/commits/dcb9ddd4d184fe4038f1745d784dc2afebc6961c)

**Review by mstechly — APPROVED** [review]

**Pull request merged**